File: src/main/java/com/example/rpn/OperatorsRegistry.java
package com.example.rpn;


import java.util.*;
import java.util.function.BinaryOperator;


public class OperatorsRegistry
{
  private Map<String,BinaryOperator<Double>> operators;

  private OperatorsRegistry(Map<String,BinaryOperator<Double>> operators)
  {
    this.operators = operators;
  }

  public static OperatorsRegistry empty()
  {
    return new OperatorsRegistry(new HashMap<>());
  }

  public OperatorsRegistry withOperator(String designation, BinaryOperator<Double> operator)
  {
    Map<String, BinaryOperator<Double>> newOperators = new HashMap<>(operators);
    newOperators.put(designation, operator);
    return new OperatorsRegistry(Collections.unmodifiableMap(newOperators));
  }

  public BinaryOperator<Double> find(String designation)
  {
    return operators.get(designation);
  }
}

File: src/main/scala/com/example/rpn/OperatorsRegistry.scala
package com.example.rpn


object OperatorsRegistry
{
  type F2[A] = (A,A) => A
  type Op = F2[Double]

  def empty = new OperatorsRegistry(Map[String,Op]())
}

import OperatorsRegistry._

class OperatorsRegistry(ops:Map[String,Op])
{
  def withOperator(designation:String, operator:Op) = 
    new OperatorsRegistry(ops + (designation -> operator))

  def find(designation:String) = ops.get(designation)
}
